Support CIDR-dependent L4 egress policies

joestringer opened the pull request:
Support the combination of `toCIDR`/`toCIDRSet` and `toPorts` in a single egress rule, so this rule will only allow traffic that matches both of these conditions. 

Current limitations:
* Up to about 14 prefix lengths may be used with this PR as-is, eg, a x/1, y/2, ... z/10. Any number of CIDRs may be configured under that set of prefix lengths. Beyond we start to hit BPF instruction limits.
  * This includes CIDRSet "exceptions", eg if you have "allow CIDR /8" and "except this /16" then it might be slightly too complex.
  * When importing a policy, if the supported number of prefix lengths is likely to be exceeded, we will reject the policy (rather than waiting until compilation time)
* Cannot select the host IP(s) via CIDR policy. If you want to filter on host IP, use "entity: host".

Related: #1684

Related dependent PRs:
* [x] #3883 (Policy refactoring)
  * [x] #3909 (Labels libraries)
* [x] #3884 (Endpoint / ipcache refactoring)
* [x] #3885 (Core BPF support)

Short-term todos:
* [x] Make sure that CIDR+L4 doesn't allow the entire CIDR as well as the specified l3+L4 range
* [x] Add cluster cidr to ipcache on startup
* [x] Improve documentation
* [x] Improve unit testing
* [x] Add basic ginkgo tests
* [x] Check the output of every tool that prints labels
  * [x] `cilium identity list`
  * [x] `cilium bpf policy get`
* [x] Fix up NPHDS `OnIPIdentityCacheChange()` to generate prefixes
  * [x] Test CIDR-dependent L7
* [x] Shift regular l3 CIDR over to this model
* [x] Test, fix IPv6 CIDRs

Longer term, unlikely to be addressed by this PR:
* LPM based BPF map for newer kernels
* Enable Ingress CIDR-dependent L4(+L7)
  * Requires minor datapath changes
  * Remove CIDRPolicyMap and related logic

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/3835)
<!-- Reviewable:end -->

joestringer committed:
ipcache: Support CIDR prefix to ID mappings

Extend the IPIdentityPair to support IP masks so that we can propagate
CIDR->ID mappings into the KVstore.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Allocate identities for CIDRs

For each unique CIDR that is configured via policy, we allocate an
Identity. These CIDR -> Identity mappings are then stored in the ipcache
in the kvstore so that we can perform LPM CIDR -> Identity mapping.

With these things in place, the core of the Cilium userspace and
datapath code can use simple IP (CIDR) -> Identity mapping, then
Identity-based policy for allowing or denying traffic to/from those
CIDRs. This greatly simplifies the handling of CIDRs throughout the
codebase.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
policy: Resolve CIDRs in rule GetAsEndpointSelectors()

This magic patch enables ingress / egress rules to select CIDRs via
EndpointSelectors. Bam. CIDR-dependent L4.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Fix ipcache conflict between hosts and prefixes

For fully specified prefixes such as ".../32" for IPv4 or ".../128" for
IPv6, the daemon's ipcache garbage collector would potentially delete
the corresponding datapath entry if there was a host in the ipcache but
not a fully specified prefix. Resolve this by attempting to look up
either of them from the garbage collector, and if either can be found,
leave the corresponding datapath entry alone.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Refactor ipcache initialization.

Shift this stuff out into its own function to shrink the size of d.init().

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
api: Allow egress CIDR+L4 rules

Also, document the behaviour.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
runtime: Add CIDR + L4/L7 egress tests.

Signed-off-by: Joe Stringer <joe@covalent.io>

ianvernon commented on pkg/policy/l4.go:
```diff
@@ -168,65 +167,40 @@ func CreateL4IngressFilter(fromEndpoints api.EndpointSelectorSlice, rule api.Por
+// CreateL4IngressFilter creates a filter for L4 policy that applies to the
+// specified endpoints and port/protocol for ingress traffic, with reference
+// to the original rules that the filter is derived from. This filter may be
+// associated wit ha series of L7 rules via the `rule` parameter.
```
wit ha --> with a

joestringer committed:
ipcache: Reject policies with too many CIDRs.

At the moment, there's a limitation in the BPF codepath where on older
kernels, the LPM implementation in use for ipcache CIDR lookups results
in many instructions generated per unique CIDR prefix length. If more
than about 16 of these are used (with ~15 instructions / prefix length),
then the BPF code we generate will exceed the kernel verification limit
for a single program, which leads endpoints to fall into a bad state.

Since we need to count the CIDRs in use anyway via the ipcache to
generate these lengths, take these limits into account when importing
the policy, and reject a policy if the resultant set of CIDRs would
generate BPF programs that are too big.

This limit is best-effort, since other things such as the number of
ports forwarded to L7 may also affect this instruction limit and it's
difficult to model the code generation ahead of time, but for basic
testing this makes it fairly obvious where the limits are for some basic
cases.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Push reserved CIDR ranges into ipcache

On startup, we need to push the cluster IP (v4 and v6) and default CIDR
mappings into the ipcache, bypassing the kvstore. We already do this for
hosts, expand it for the other special identities.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
daemon: Populate BPF ipcache with CIDR prefixes

Implement prefix length tracking in the ipcache so that the daemon can
request the list of prefixes during BPF code generation.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer commented on pkg/ipcache/ipcache.go:
```diff
@@ -108,18 +124,49 @@ func UpsertIPToKVStore(IP net.IP, ID identity.NumericIdentity, metadata string)
+
+// UpsertIPNet updates / inserts the provided CIDR->Identity mapping into the
+// kvstore, which will subsequently trigger an event in ipIdentityWatcher().
+func UpsertIPNet(prefix *net.IPNet, ID *identity.Identity) error {
```
`s/UpsertIPNet/UpsertIPNetToKVStore/`

joestringer commented on pkg/ipcache/ipcache.go:
```diff
@@ -108,18 +124,49 @@ func UpsertIPToKVStore(IP net.IP, ID identity.NumericIdentity, metadata string)
+//
+// If any Prefix->Identity mapping cannot be created, it will not create any
+// of the mappings and returns an error.
+func UpsertIPNets(prefixes []*net.IPNet, identities []*identity.Identity) (err error) {
```
`s/UpsertIPNets/UpsertIPNetsToKVStore/`

joestringer commented on pkg/ipcache/ipcache.go:
```diff
@@ -129,6 +176,22 @@ func DeleteIPFromKVStore(ip string) error {
+// DeleteIPNets removes the Prefix->Identity mappings for the specified slice
+// of prefixes from the kvstore, which will subsequently trigger an event in
+// ipIdentityWatcher().
+func DeleteIPNets(prefixes []*net.IPNet) (err error) {
```
`s/DeleteIPNets/DeleteIPNetsFromKVStore/`

joestringer commented on daemon/policy.go:
```diff
@@ -222,8 +225,28 @@ func (d *Daemon) policyAdd(rules api.Rules, opts *AddOptions) (uint64, error) {
+	prefixIdentities, err := identity.AllocateCIDRIdentities(prefixes)
+	if err != nil {
+		metrics.PolicyImportErrors.Inc()
+		return d.policy.GetRevision(), err
```
We should probably shift this CIDR Identity resolution logic to a controller. Could be a follow-up PR.

joestringer replied:
Filed this issue to follow up: https://github.com/cilium/cilium/issues/3942

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -951,22 +952,40 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+					break
+				}
+			}
+			Expect(otherHostIP).To(Not(Equal("")))
```
`Expect(otherHostIP).NotTo(Equal(""))`

I prefer to use `NotTo` is a bit simpler IMHO.

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -951,22 +952,40 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+				}
+			}
+			Expect(otherHostIP).To(Not(Equal("")))
+			By(fmt.Sprintf("Using %s for world CIDR IP", otherHostIP))
```
I would suggest to delete this one, `otherHostIP` is in other 3 debug messages, don't add a real value for me this one ;-)

joestringer replied:
This message is stating what the IP address represents, the other logs don't say that. I think it would be easier when debugging to see something like:

* Detecting host IP in world CIDR
* Using 192.0.2.3 for world CIDR IP
* Pinging 192.0.02.3 from app1 before importing policy
* ...

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -951,22 +952,40 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
 
 		BeforeEach(func() {
-			By(fmt.Sprintf("Pinging %s from %s before importing policy (should work)", api.EntityHost, helpers.App1))
+			By(fmt.Sprintf("Pinging %s from %s before importing policy (should work)", hostIP, helpers.App1))
```
On `By` messages I would suggest the use of `%q` so the variables are into `''` and it's easy to understand what are the moving parts ;-)

joestringer replied:
I assume it's always the same for strings, but for objects this is better? OK :)

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1015,6 +1032,118 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+
+			By(fmt.Sprintf("Accessing /private on %s from %s should fail", otherHostIP, helpers.App1))
+			res := vm.ContainerExec(helpers.App1, helpers.CurlWithHTTPCode("http://%s/private", otherHostIP))
+			res.ExpectContains("403", "unexpectedly able to access http://%s:80/private when access should only be allowed to /private", otherHostIP)
```
This assert message does not make sense to me no? should be: `unexpectedly able to access http://%s:80/private when access should only be allowed to /index.html", otherHostIP` no?

eloycoto reviewed (commented):
Few minor things related to test ;-)

ianvernon commented on pkg/ipcache/ipcache.go:
```diff
@@ -226,20 +227,49 @@ const (
 )
 
-func keyToIP(key string) (net.IP, error) {
+// keyToIPNet returns the IPnet describing the key, whether it is a host, and
```
Nit: IPnet --> IPNet

ianvernon commented on pkg/ipcache/ipcache.go:
```diff
@@ -263,60 +293,102 @@ func ipIdentityWatcher(listeners []IPIdentityMappingListener) {
 			if cacheChanged {
 				log.WithFields(logrus.Fields{
-					"endpoint-ip":          ipIDPair.IP,
+					logfields.IPAddr:       ipIDPair.IP,
```
As part of this PR, add yourself as a codeowner in `CODEOWNERS` given the amount of changes done to ipcache ;)

joestringer replied:
Sure. Shall we create a team for it?

ianvernon commented on daemon/policy.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2016-2017 Authors of Cilium
+// Copyright 2016-2018 Authors of Cilium
```
Typo in commit message: ` we can perform LPM CIDR ->m Identity mapping.` should be ` we can perform LPM CIDR -> Identity mapping.`

ianvernon commented on daemon/policy.go:
```diff
@@ -222,8 +225,28 @@ func (d *Daemon) policyAdd(rules api.Rules, opts *AddOptions) (uint64, error) {
 	log.WithField(logfields.CiliumNetworkPolicy, logfields.Repr(rules)).Debug("Policy Add Request")
 
+	prefixes := policy.GetCIDRPrefixes(rules)
+	log.WithField("prefixes", prefixes).Debug("Policy imported via API, found prefixes...")
```
"found prefixes" --> "found CIDR prefixes" ?

ianvernon commented on pkg/ipcache/ipcache.go:
```diff
@@ -129,6 +177,22 @@ func DeleteIPFromKVStore(ip string) error {
+		}
+	}
+
+	return err
```
`return err` --> `return`

ianvernon commented on pkg/ipcache/ipcache.go:
```diff
@@ -327,6 +327,34 @@ func (ipc *IPCache) LookupByIPRLocked(IP string) (identity.NumericIdentity, bool
 }
 
+// LookupByPrefixRLocked looks for either the specified CIDR prefix, or if the
+// prefix is fully specified (ie, w.x.y.z/32 for IPv4), find the host for the
```
find --> finds

ianvernon commented on Documentation/policy/language.rst:
```diff
@@ -496,6 +496,32 @@ endpoints with the label ``role=frontend`` will not be able to communicate with
 
         .. literalinclude:: ../../examples/policies/l4/l3_l4_combined.json
 
+CIDR-dependent Layer 4 rule
```
rule --> Rule

ianvernon commented on pkg/ipcache/ipcache.go:
```diff
@@ -159,7 +197,45 @@ func (ipc *IPCache) deleteLocked(IP string) {
+
+// ToBPFData renders the ipcache into the relevant set of CIDR prefixes for
+// the BPF datapath to use for lookup.
+func (ipc *IPCache) ToBPFData() (s6, s4 []int) {
```
The ipcache should be locked here.

joestringer replied:
Good spotting, I completely missed this.

ianvernon reviewed (commented):
All I have is nitpicks for the most part. The code seems sound to me, but I probably will give it another review since it is large in size and I want to give it some time to digest.

joestringer committed:
CODEOWNERS: Shift ownership of ipcache to a team

Signed-off-by: Joe Stringer <joe@covalent.io>

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -951,22 +951,40 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+			res.ExpectSuccess("No docker bridge available for testing egress CIDR within host")
+			filter := fmt.Sprintf(`{ [0].IPAM.Config[0].Gateway }`)
+			obj, err := res.FindResults(filter)
+			Expect(err).NotTo(HaveOccurred())
```
Assert message

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -951,22 +951,40 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+			filter := fmt.Sprintf(`{ [0].IPAM.Config[0].Gateway }`)
+			obj, err := res.FindResults(filter)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(obj).To(HaveLen(1))
```
assert message

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -951,22 +951,40 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+			Expect(obj).To(HaveLen(1))
+			otherHostIP = obj[0].Interface().(string)
+			Expect(otherHostIP).Should(MatchRegexp("^[.:0-9a-f][.:0-9a-f]*$"))
+			By(fmt.Sprintf("Using %s for world CIDR IP", otherHostIP))
```
Please, if you want to use this BY please use `%q` please.

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1015,6 +1031,119 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+		// for it.
+		testCIDRL4Policy := func(policy, dstIP, proto string) {
+			_, err := vm.PolicyRenderAndImport(policy)
+			ExpectWithOffset(1, err).To(BeNil(), "Unable to import policy: %s", err)
```
This `err` in the assert message is redundant, the `err` message is already in the assert.

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1015,6 +1031,119 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+			_, err := vm.PolicyRenderAndImport(policy)
+			ExpectWithOffset(1, err).To(BeNil(), "Unable to import policy: %s", err)
+
+			By(fmt.Sprintf("Pinging %s from %s (should not work)", api.EntityHost, helpers.App1))
```
use %q please

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1015,6 +1031,119 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+			By(fmt.Sprintf("Pinging %s from %s (should not work)", api.EntityHost, helpers.App1))
+			res := vm.ContainerExec(helpers.App1, helpers.Ping(dstIP))
+			ExpectWithOffset(1, res.WasSuccessful()).Should(BeFalse(),
+				"expected ping to %s to fail", dstIP)
```
use %q please

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1015,6 +1031,119 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+
+			// Docker container running with host networking is accessible via
+			// the docker bridge's IP address. See https://docs.docker.com/network/host/.
+			By(fmt.Sprintf("Accessing index.html using Docker container using host networking from %s (should work)", helpers.App1))
```
use %q please

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1015,6 +1031,119 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+			ExpectWithOffset(1, res.Output().String()).To(ContainSubstring("200"),
+				"Expected to be able to access /public in host Docker container")
+
+			By(fmt.Sprintf("Accessing %s on wrong port from %s should fail", dstIP, helpers.App1))
```
use %q please

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1015,6 +1031,119 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+			ExpectWithOffset(1, res.WasSuccessful()).Should(BeFalse(),
+				"unexpectedly able to access %s when access should only be allowed to CIDR", dstIP)
+
+			By(fmt.Sprintf("Accessing port 80 on wrong destination from %s should fail", helpers.App1))
```
use %q please

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1015,6 +1031,119 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+			ExpectWithOffset(1, res.WasSuccessful()).Should(BeFalse(),
+				"unexpectedly able to access %s when access should only be allowed to CIDR", hostIP)
+
+			By(fmt.Sprintf("Pinging %s from %s (shouldn't work)", helpers.App2, helpers.App1))
```
use %q please

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1015,6 +1031,119 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+			ExpectWithOffset(1, err).Should(BeNil(),
+				"Unable to get networking information for container %s", helpers.Httpd2)
+
+			By(fmt.Sprintf("Accessing /index.html in %s from %s (shouldn't work)", helpers.App2, helpers.App1))
```
use %q please

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1015,6 +1031,119 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+		}
+
+		It("Tests egress with CIDR+L4 policy", func() {
+			By(fmt.Sprintf("Importing policy which allows egress to %s from %s", otherHostIP, helpers.App1))
```
use %q

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -1015,6 +1031,119 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+		It("Tests egress with CIDR+L4 policy to external https service", func() {
+			cloudFlare := "1.1.1.1"
+
+			By(fmt.Sprintf("Importing policy which allows egress to %s from %s", otherHostIP, helpers.App1))
```
Please use %q in all BY messages. It's easy to found the moving parts.

raybejjani commented on Documentation/policy/language.rst:
```diff
@@ -496,6 +496,32 @@ endpoints with the label ``role=frontend`` will not be able to communicate with
+they must communicate using TCP on port 80. Endpoints with other labels will
+not be able to communicate with destinations in the CIDR ``192.0.2.0/24``, and
+endpoints with the label ``role=crawler`` will not be able to communicate with
+destinations in the CIDR ``192.0.2.0/24`` on ports other than 80.
```
`and endpoints with the label ``role=crawler`` will not be able to communicate with destinations in the CIDR ``192.0.2.0/24`` on ports other than 80.` is a little redundant with the first sentence.

Endpoint with other labels is a little unclear. `Endpoints without the label role=crawler will not..` is more direct about this policy not affecting other endpoints (I'm not sure if we should be more direct and say `The policy of endpoints without label=crawler is not affected).

joestringer replied:
The redundancy is intentional, to clarify exactly that the policy only applies to traffic which matches the intersection of the CIDR and L4, and not the union. (Perhaps if there's some better wording I could just say this... hmm)

I've updated the latter point though, thanks.

raybejjani commented on daemon/daemon.go:
```diff
@@ -906,31 +913,67 @@ func (d *Daemon) syncLXCMap() error {
+				return fmt.Errorf("Unable to add host entry to endpoint map: %s", err)
+			}
+			if added {
+				log.WithField(logfields.IPAddr, pair.IP).Debugf("Adding local ip to endpoint map")
```
"Added" instead of "Adding".

raybejjani commented on pkg/identity/identity.go:
```diff
@@ -98,3 +101,20 @@ func (id *Identity) GetModel() *models.Identity {
+// PrefixString returns the IPIdentityPair's IP as either a host IP in the
+// format w.x.y.z if 'host' is true, or as a prefix in the format the w.x.y.z/N
+// if 'host' is false.
+func (pair *IPIdentityPair) PrefixString(host bool) string {
```
Do we need the `host` parameter? I found it confusing since we triggered it slightly different ways when we used it:
1- `isHost := pair.ID == identity.ReservedIdentityHost`
2- `isHost := ipIDPair.Mask == nil`

but (2) is always correct when generating the string (and if (1) is a special case then that is sort of ok since this is all inside the identity package). It seems weird to me that the caller needs to figure out if this is a CIDR range or not. If it can't be stored in the IPIdentityPair itself then do we just need 2 separate classes?

joestringer replied:
Yeah this is a bit ugly, I haven't spent any time trying to clean it up. You're right that (1) is a special case, though in theory this could be handled by setting the mask to nil. For (2), there are a couple of other special cases. The delete case is simple and we don't require the parameter there. The other case is in `findShadowedCIDR`, where it's possible to have both `192.0.2.3` and `192.0.2.3/32`. In this case, the identity for the host entry is prioritized. Both could exist in the kvstore at once, and when the host is deleted, then the ipcache needs to swap out the host's identity with the CIDR prefix's identity. To detect this case, we need to be able to turn the host entry into a longest-possible-match prefix for the protocol in use (IPv4/IPv6) and perform a lookup with that string. We could do a bit of special code for that case instead of having this parameter though.

joestringer replied:
@raybejjani I added a new commit on top for this.

raybejjani commented on pkg/identity/allocator.go:
```diff
@@ -120,3 +120,19 @@ func AllocateIdentity(lbls labels.Labels) (*Identity, bool, error) {
 }
+
+// ReleaseSlice attempts to release a set of identities. It is a helper
+// function that may be useful for cleaning up multiple identitiese in paths
```
typo `identitiese`

raybejjani commented on pkg/identity/cidr.go:
```diff
@@ -0,0 +1,87 @@
+			continue
+		}
+		labels := cidr.GetCIDRLabels(prefix)
+		id, _, err := AllocateIdentity(labels)
```
Don't we need to track `isNew` here? (the _ return value). Overlapping CIDRs can generate the same identity (one with a more specific range, I think) but then, when we clean up on error below, we would try to release it even if the allocate was a no-op in the KVStore (thus releasing that ID globally).

raybejjani replied:
Oh, the kvstore code is refcounting for us. This is probably fine.

joestringer replied:
Yeah, kvstore handles this.

joestringer replied:
Oh, and overlapping CIDRs have different sets of labels, so they have different identities.

raybejjani commented on pkg/identity/cidr.go:
```diff
@@ -0,0 +1,87 @@
+// On error, returns all identities that can be resolved and an error.
+func LookupCIDRIdentities(prefixes []*net.IPNet) (res []*Identity, err error) {
+	res = make([]*Identity, 0, len(prefixes))
+	err = nil
```
since this is a named return variable, it's already nil (the 0 value for an `error`)

joestringer replied:
While debugging at one point, I was beginning to doubt this.. I'll fix it up :-)

raybejjani commented on pkg/identity/cidr.go:
```diff
@@ -0,0 +1,87 @@
+		}
+	}
+
+	return res, err
```
It's unfortunate that we only return the most recent error here. Similarly, we don't associate the error with the identity we were trying to get. I'm not sure it matters, though, since the errors here should be network type things, right? not anything specific to an identity?

joestringer replied:
I believe that it shouldn't be anything to do with the identity, I've tested with a variety of different generated sets of labels including ones for prefixes like "::/128".

The assumption here is very much that if there's one problem, there may be many problems, and they're unlikely to differ.

Perhaps when I look at putting a controller in charge here, we could be a bit more thorough in our error reporting.

raybejjani commented on pkg/identity/cidr.go:
```diff
@@ -0,0 +1,87 @@
+		log.WithError(err).Warningf("Failed to allocate identities for %d prefixes", len(res))
+	}
+
+	return res, err
```
It might be worth returning which identities could not be allocated? I imagine the causes here could be more varied and may be specific to the identity? (I can't think of anything, though, so maybe not)

joestringer replied:
I suppose that if we exit out & try to recover / clean up on the first failure, we could return the prefix and/or labels for that first failure, then log these in the caller. Then if there really is some problem of that variety, it should be easy to tell.

raybejjani replied:
your call. I think I came to this from the perspective of overlapping ranges sharing some identities. You corrected me above, so that leaves classes of errors that are likely (as you said) going to repeat for all of them or none of them.

raybejjani commented on daemon/daemon.go:
```diff
@@ -123,6 +123,9 @@ type Daemon struct {
 	compilationMutex *lock.RWMutex
+
+	// ipcacheListeners lists all parties interested in IP -> ID mappings.
+	ipcacheListeners []ipcache.IPIdentityMappingListener
```
This is a little out of scope for this PR but could we move this to the ipcache package? It's a singleton that's private to that package except we want to notify the listeners manually. We can have a function to do just that as part of the package (as weird as I find the whole thing).
One day, when someone wants to add a new listener they won't realise that this is a constant and they will change it.

joestringer replied:
Sure. #3976

raybejjani replied:
Thanks :)

joestringer committed:
identity: Fix pair.PrefixString() arguments

Previously, the arguments forced the callers to be aware of the details
of how the IPIdentityPair represents a host->Identity mapping. This
commit hides those details within the implementation, then provides an
accessor to determine whether this is the case, for users that care.

Signed-off-by: Joe Stringer <joe@covalent.io>

eloycoto commented on test/runtime/Policies.go:
```diff
@@ -951,21 +951,39 @@ var _ = Describe("RuntimeValidatedPolicies", func() {
+			Expect(err).NotTo(HaveOccurred(), "Error occurred while finding docker bridge IP")
+			Expect(obj).To(HaveLen(1), "Unexpectedly found more than one IPAM config element for docker bridge")
+			otherHostIP = obj[0].Interface().(string)
+			Expect(otherHostIP).Should(MatchRegexp("^[.:0-9a-f][.:0-9a-f]*$"))
```
Here the assert message

joestringer replied:
Ah shoot, I thought I caught them all. I'll fix this up, thanks.

eloycoto reviewed (commented):
Minor nitpick

jrfastab reviewed (approved):
Looks reasonable to me.
